agent/consul/acl_endpoint.go

```diff
@@ -2,239 +2,866 @@ package consul
 
 import (
 	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
 	"time"
 
 	"github.com/armon/go-metrics"
 	"github.com/hashicorp/consul/acl"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/lib"
 	"github.com/hashicorp/go-memdb"
 	"github.com/hashicorp/go-uuid"
 )
 
+const (
+	// aclBootstrapReset is the file name to create in the data dir. It's only contents
+	// should be the reset index
+	aclBootstrapReset = "acl-bootstrap-reset"
+)
+
+// Regex for matching
+var validPolicyName = regexp.MustCompile(`^[A-Za-z0-9\-_]{1,128}$`)
+
 // ACL endpoint is used to manipulate ACLs
 type ACL struct {
 	srv *Server
 }
 
+func (a *ACL) fileBootstrapResetIndex() uint64 {
+	// Determine the file path to check
+	path := filepath.Join(a.srv.config.DataDir, aclBootstrapReset)
+
+	// Read the file
+	raw, err := ioutil.ReadFile(path)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			a.srv.logger.Printf("[ERR] acl.bootstrap: failed to read %q: %v", path, err)
+		}
+		return 0
+	}
+
+	// Attempt to parse the file
+	var resetIdx uint64
+	if _, err := fmt.Sscanf(string(raw), "%d", &resetIdx); err != nil {
+		a.srv.logger.Printf("[ERR] acl.bootstrap: failed to parse %q: %v", path, err)
+		return 0
+	}
+
+	// Return the reset index
+	a.srv.logger.Printf("[WARN] acl.bootstrap: parsed %q: reset index %d", path, resetIdx)
+	return resetIdx
+}
+
+func (a *ACL) aclPreCheck() error {
+	if !a.srv.ACLsEnabled() {
+		return acl.ErrDisabled
+	}
+
+	if a.srv.UseLegacyACLs() {
+		return fmt.Errorf("The ACL system is currently in legacy mode.")
+	}
+
+	return nil
+}
+
 // Bootstrap is used to perform a one-time ACL bootstrap operation on
 // a cluster to get the first management token.
-func (a *ACL) Bootstrap(args *structs.DCSpecificRequest, reply *structs.ACL) error {
+func (a *ACL) Bootstrap(args *structs.DCSpecificRequest, reply *structs.ACLToken) error {
+	if err := a.aclPreCheck(); err != nil {
+		return err
+	}
 	if done, err := a.srv.forward("ACL.Bootstrap", args, args, reply); done {
 		return err
 	}
 
 	// Verify we are allowed to serve this request
-	if a.srv.config.ACLDatacenter != a.srv.config.Datacenter {
+	if !a.srv.InACLDatacenter() {
 		return acl.ErrDisabled
 	}
 
 	// By doing some pre-checks we can head off later bootstrap attempts
 	// without having to run them through Raft, which should curb abuse.
 	state := a.srv.fsm.State()
-	bs, err := state.ACLGetBootstrap()
+	allowed, resetIdx, err := state.CanBootstrapACLToken()
 	if err != nil {
 		return err
 	}
-	if bs == nil {
-		return structs.ACLBootstrapNotInitializedErr
-	}
-	if !bs.AllowBootstrap {
-		return structs.ACLBootstrapNotAllowedErr
+	var specifiedIndex uint64 = 0
+	if !allowed {
+		// Check if there is a reset index specified
+		specifiedIndex = a.fileBootstrapResetIndex()
```

Comment:
It is an override to allow bootstrapping again and is allowed if the integer in the file matches the value within the index memdb table for "acl-token-bootstrap" key.

Why not config? Config is considered readable and we need to be able to blow this file away when successful so that we don't accidentally allow the override if a snapshot is restored.

Also we never write to this file. An admin with shell access on the node would have to write the file manually. Consul just reads and later unlinks it.